File: java-gdb-activities-20/activity18/src/com/gdb/logging/DatabaseLogDestination.java
package com.gdb.logging;

import com.gdb.command.TransactionCommand;
import com.gdb.db.SimulatedDatabase;
import java.util.*;

public class DatabaseLogDestination implements LogDestination {
    private static final String TABLE = "transaction_log";

    // ============================================================
    // 📝 STEP 11: Declare Field
    // ============================================================
    // TODO: declare private final SimulatedDatabase db;

    // ============================================================
    // 📝 STEP 12: Constructor
    // ============================================================
    // TODO: implement constructor accepting SimulatedDatabase
    public DatabaseLogDestination(SimulatedDatabase db) {
        // TODO: Step 12 - implement constructor
    }

    // ============================================================
    // 📝 STEP 13: write(cmd)
    // ============================================================
    // TODO: insert command into database table
    @Override
    public void write(TransactionCommand cmd) {
        // TODO: Step 13 - insert command into database table
    }

    // ============================================================
    // 📝 STEP 14: readAll()
    //
    // INSTRUCTIONS:
    //   1. Call db.selectAll(TABLE).
    //   2. Map/cast each Object to TransactionCommand.
    //   3. Collect and return List<TransactionCommand>.
    // ============================================================
    // TODO: retrieve and return all commands from database
    @Override
    public List<TransactionCommand> readAll() {
        // TODO: Step 14 - retrieve and return all commands from database
        return new ArrayList<>();
    }

    // ============================================================
    // 📝 STEP 15: clear()
    // ============================================================
    // TODO: delete all records from database table
    @Override
    public void clear() {
        // TODO: Step 15 - delete all records from database table
    }

    // ============================================================
    // 📝 STEP 16: getDestinationName()
    // ============================================================
    // TODO: return "DATABASE"
    @Override
    public String getDestinationName() {
        // TODO: Step 16 - return "DATABASE"
        return "";
    }
}

File: python-gdb-activities-20/activity18/gdb/logging/database_log_destination.py
# gdb/logging/database_log_destination.py
from typing import List
from gdb.domain.transaction import Transaction
from gdb.logging.log_destination import LogDestination
from gdb.db.simulated_database import SimulatedDatabase

class DatabaseLogDestination(LogDestination):
    """Database storage destination implementor backed by SimulatedDatabase."""

    # ============================================================
    # 📝 STEP 3: Implement DatabaseLogDestination
    #
    # INSTRUCTIONS:
    #   1. __init__(db): store the SimulatedDatabase in self._db.
    #   2. write(transaction): insert the transaction into the database.
    #   3. read_all(): return every transaction stored in the database.
    #   4. get_destination_name(): return "Database".
    #
    # HINT: SimulatedDatabase (gdb/db/simulated_database.py) is complete -- read its insert()
    #       and query_all() methods first.
    # ============================================================
    # TODO: keep a reference to the database
    def __init__(self, db: SimulatedDatabase) -> None:
        raise NotImplementedError("TODO: Step 3 - implement DatabaseLogDestination constructor")

    def write(self, transaction: Transaction) -> None:
        # TODO: insert the transaction into the database
        raise NotImplementedError("TODO: Step 3 - implement write")

    def read_all(self) -> List[Transaction]:
        # TODO: return all transactions stored in the database
        raise NotImplementedError("TODO: Step 3 - implement read_all")

    def get_destination_name(self) -> str:
        # TODO: return "Database"
        raise NotImplementedError("TODO: Step 3 - implement get_destination_name")
